plugin-server/src/worker/ingestion/team-manager.ts

```diff
@@ -39,7 +48,45 @@ export class TeamManager {
         }
     }
 
-    public async updateEventNamesAndProperties(teamId: number, event: string, properties: Properties): Promise<void> {
+    async flushLastSeenAtCache(): Promise<void> {
+        let valuesStatement = ''
+        let params: (string | number)[] = []
+
+        const events = this.eventLastSeenCache
+        this.eventLastSeenCache = new Map()
+
+        for (const event of events) {
+            const [key, value] = event
+            const teamId = key.split('_', 1)[0]
+            const eventName = key.substring(key.indexOf(teamId) + teamId.length + 1)
+            if (teamId && eventName && value) {
+                valuesStatement += `($${params.length + 1}, $${params.length + 2}, $${params.length + 3}),`
+                params = params.concat([teamId, eventName, value])
+            }
+        }
+
+        this.lastFlushAt = DateTime.now()
+
+        if (params.length) {
+            await this.db.postgresQuery(
+                `UPDATE posthog_eventdefinition AS t1 SET last_seen_at = GREATEST(t1.last_seen_at, to_timestamp(t2.last_seen_at::integer))
+                FROM (VALUES ${valuesStatement.substring(
+                    0,
+                    valuesStatement.length - 1
+                )}) AS t2(team_id, name, last_seen_at)
```

Comment:
This `valuesStatement.substring(0,valuesStatement.length - 1)` feels like something can go horribly wrong. Just as a practice we should almost never really inject sql into sql, and removing a comma like that is... smelly 🐟 

It's also not clear why `from (values())`, and not just something like:

```sql
UPDATE posthog_eventdefinition SET last_seen_at = GREATEST(last_seen_at, $1), created_at = LOWEST(created_at, $2) where team_id = $3 and name = $4
```

Finally, I'd check about running all of this in a transaction. I can't 100% confirm this, but I think a transaction will be sent to postgres in one go. So if we have 1000 events to update, here we're awaiting for 1000 queries to return, so combining them all into one block that's sent to the backend at the same time will be faster.

Reply:
Alright, so the reason this uses `from (values())` and we inject SQL (which doesn't contain any user input so should be clean), is to run the update statement in one go, it's a single query. I'll switch to transaction-based updates in my last commit here, so you can compare and advise on which option would be best.